### PR TITLE
make x[...] .= ... assign in-place

### PR DESCRIPTION
**Pull request by stevengj**
This fixes a bug in #17510: `x[...] .= ...` and `x[...] .+= ...` etcetera should call `broadcast!` on `view(x, ...)`, so that it continues to work in-place.   (In Julia 0.4, `x[...] .+= ...` translates into a `setindex!` call.)

### CODE REVIEW

**Comment by stevengj** [comment]
(@timholy, is there anything about the current `view` implementation that would prevent `broadcast!(identity, view(x, ...), rhs)` from being a drop-in replacement for `setindex!(x, rhs, ...)`?  It seems to work in all the cases I've tried.)

File: src/julia-syntax.scm [hunk 1]
```diff
@@ -1542,6 +1542,15 @@
         (cadr expr)  ;; eta reduce `x->f(x)` => `f`
         `(-> ,argname (block ,@splat ,expr)))))
 
+(define (ref-to-view expr)
+  (if (and (pair? expr) (eq? (car expr) 'ref))
+      (let* ((ex (partially-expand-ref expr))
+             (stmts (butlast (cdr ex)))
+             (refex (last    (cdr ex)))
+             (nuref `(call view ,(caddr refex) ,@(cdddr refex))))
```
Comment (stevengj) [inline src/julia-syntax.scm]: @JeffBezanson, should this be `(top view)` rather than `view`?

File: src/julia-syntax.scm [hunk 2]
```diff
@@ -1657,14 +1666,15 @@
...
+        (lhs_ (ref-to-view lhs))) ; x[...] expressions on lhs turn in to view(x, ...) to update x in-place
     (if (fuse? e)
         (if (null? lhs)
             (expand-forms `(call broadcast ,(from-lambda (cadr e)) ,@(caddr e)))
-            (expand-forms `(call broadcast! ,(from-lambda (cadr e)) ,lhs ,@(caddr e))))
+            (expand-forms `(call broadcast! ,(from-lambda (cadr e)) ,lhs_ ,@(caddr e))))
         (if (null? lhs)
             (expand-forms e)
-            (expand-forms `(call broadcast! identity ,lhs ,e))))))
+            (expand-forms `(call broadcast! identity ,lhs_ ,e))))))
```
Comment (stevengj) [inline src/julia-syntax.scm]: Similarly, should these be calls to `(top broadcast)` and `(top broadcast!)`?

**Comment by stevengj** [comment]
Marking as 0.5, since this needs to be resolved one way or the other (either by merging this in some form or by reverting #17510) by 0.5.

File: src/julia-syntax.scm (hunk 1 above)
Comment (JeffBezanson) [inline src/julia-syntax.scm]: Yes; I would go with `top` to avoid capturing locals that might be called `view` or `broadcast`.

**Comment by stevengj** [comment]
Okay to merge?

**Comment by tkelman** [comment]
would be good to get a confirmation on https://github.com/JuliaLang/julia/pull/17546#issuecomment-234424832 ?

**Comment by stevengj** [comment]
@tkelman, even if there is some case where `view` is problematic, I think that this PR is a strict improvement on the broken behavior we have currently (my fault!).

If `view` fails, my suspicion is that it will fail noisily — most likely a `MethodError`, if there is some custom array-like type  that implements an odd `getindex` method but not a corresponding `view` method.   (e.g. PyCall uses `getindex` in a weird way, and `view` is not implemented for PyObject.  I'll have to think about whether `.=` even makes sense for a `PyObject`.)

**Comment by StefanKarpinski** [comment]
Seems like sound reasoning, @stevengj  – in favor of merging.

**Comment by stevengj** [comment]
One case that we can't handle yet is that of a dictionary.   Neither `view` nor `broadcast!` currently handle dictionaries, so something like `mydict[key] .*= ...` will throw a `MethodError` for now.

One way to handle this would be to implement a `view(a::Associative, key)` method that just returns `a[key]`, I think.   That way, if `a[key]` is a mutable array then `a[key] .= ...` will work as expected, while if `a[key]` is an immutable object then `broadcast!` will throw a `MethodError`.

**Comment by iamed2** [comment]
I think people are going to be confused when `.=` works when LHS is an array but not when LHS returns an array. Another case is:
```julia
julia> a = [[4, 5], [6, 7]]
julia> a[1] .= 3
ERROR: MethodError: Cannot `convert` an object of type Int64 to an object of type Array{Int64,1}
This may have arisen from a call to the constructor Array{Int64,1}(...),
since type constructors fall back to convert methods.
 in fill!(::SubArray{Array{Int64,1},0,Array{Array{Int64,1},1},Tuple{Int64},false}, ::Int64) at ./multidimensional.jl:549
 in broadcast!(::Base.#identity, ::SubArray{Array{Int64,1},0,Array{Array{Int64,1},1},Tuple{Int64},false}, ::Int64) at ./broadcast.jl:19
 in eval(::Module, ::Any) at ./boot.jl:234
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:46
```
The way to get the intended result is:
```julia
julia> a[1][:] .= 3
2-element SubArray{Int64,1,Array{Int64,1},Tuple{Colon},true}:
 3
 3
julia> a
2-element Array{Array{Int64,1},1}:
 [3,3]
 [6,7]
```
but that's not exactly obvious and might require a manual entry (unless a special case is made for scalar indexing). 

AFAIK this is why MATLAB cell arrays have two types of indexing brackets: `()` to slice and `{}` to extract (not that I'm suggesting that).

**Comment by stevengj** [comment]
@iamed, I think that most cases of interest could be supported by adding appropriate methods to `broadcast!` and/or `view`.

e.g. your particular case could be handled by defining:
```jl
Base.broadcast!{T<:AbstractArray}(f, x::SubArray{T,0}, args...) = broadcast!(f, x[], args...)
Base.broadcast!{T<:AbstractArray}(f::typeof(identity), x::SubArray{T,0}, arg::Number) = broadcast!(f, x[], arg) # prevent method ambiguity
```

We could also get more fine-grained control by defining a `broadcastview(...)` function, which defaults to `view(...)`, to allow us to define behaviors that only affect the implicit views `.=` and not explict calls to `view`.

**Comment by stevengj** [comment]
Note also that this PR can be reverted in 0.6 if the arraypocolypse happens and slices become views.   So potentially this is just a temporary fix.

**Pull request merged**